From fe09fac168cfb5930e7ab9a921f3acc4be0bc32e Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Sun, 16 May 2021 10:12:32 -0700
Subject: [PATCH 100/196] Implement XAddress parsing (#173)

Adds a `Parse` method to the XAddress namespace that uses the
XAddress Decode operation and then feeds it into `ExtractDestination`.
This returns a CTxDestination which is usable in the rest of the code
base and wallet.
---
 src/addresses/xaddress.cpp  | 37 ++++++++++++++
 src/addresses/xaddress.h    |  9 ++++
 src/test/xaddress_tests.cpp | 99 +++++++++++++++++++++++++++++++++++++
 3 files changed, 145 insertions(+)

diff --git a/src/addresses/xaddress.cpp b/src/addresses/xaddress.cpp
index df7bcf25c..6e2d8afbe 100644
--- a/src/addresses/xaddress.cpp
+++ b/src/addresses/xaddress.cpp
@@ -4,12 +4,16 @@
 
 #include <addresses/xaddress.h>
 #include <base58.h>
+#include <chainparams.h>
 #include <hash.h>
+#include <script/script.h>
+#include <script/standard.h>
 #include <uint256.h>
 #include <util/strencodings.h>
 #include <util/string.h>
 #include <util/vector.h>
 
+#include <assert.h>
 #include <string>
 
 namespace XAddress {
@@ -81,4 +85,37 @@ DecodeError Decode(const std::string &address, Content &parsedOutput) {
     return DECODE_OK;
 }
 
+static NetworkType GetAddressNetworkByte(const CChainParams &p) {
+    if (p.NetworkIDString() == CBaseChainParams::MAIN) {
+        return MAINNET;
+    }
+    if (p.NetworkIDString() == CBaseChainParams::TESTNET) {
+        return TESTNET;
+    }
+    if (p.NetworkIDString() == CBaseChainParams::REGTEST) {
+        return REGTEST;
+    }
+    assert(false || "Unknown network for XAddress");
+    return UNKNOWN;
+}
+
+bool Parse(const CChainParams &params, const std::string &address,
+           CTxDestination &retDestination) {
+    Content content;
+    if (Decode(address, content) != DECODE_OK) {
+        return false;
+    }
+    if (GetAddressNetworkByte(params) != content.m_network) {
+        return false;
+    }
+    if (TOKEN_NAME != content.m_token) {
+        return false;
+    }
+    if (SCRIPT_PUB_KEY != content.m_type) {
+        return false;
+    }
+    CScript scriptPubKey(content.m_payload.begin(), content.m_payload.end());
+    return ExtractDestination(scriptPubKey, retDestination);
+}
+
 } // namespace XAddress
diff --git a/src/addresses/xaddress.h b/src/addresses/xaddress.h
index 1e0f6055a..6933588b0 100644
--- a/src/addresses/xaddress.h
+++ b/src/addresses/xaddress.h
@@ -5,6 +5,9 @@
 #ifndef BITCOIN_XADDR_H
 #define BITCOIN_XADDR_H
 
+#include <chainparams.h>
+#include <script/standard.h>
+
 #include <string>
 #include <vector>
 
@@ -16,6 +19,7 @@ enum NetworkType : uint8_t {
     MAINNET = '_',
     TESTNET = 'T',
     REGTEST = 'R',
+    UNKNOWN = 'U'
 };
 
 enum AddressType : uint8_t {
@@ -54,6 +58,11 @@ std::string Encode(const Content &addressContent);
  * modified.
  */
 DecodeError Decode(const std::string &address, Content &parsedOutput);
+/**
+ * Parse an XAddress string into a CTxDestination. Returns false on failure.
+ */
+bool Parse(const CChainParams &params, const std::string &address,
+           CTxDestination &retDestination);
 } // namespace XAddress
 
 #endif // BITCOIN_XADDR_H
diff --git a/src/test/xaddress_tests.cpp b/src/test/xaddress_tests.cpp
index 4599acaea..a2b8c4368 100644
--- a/src/test/xaddress_tests.cpp
+++ b/src/test/xaddress_tests.cpp
@@ -3,7 +3,10 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <addresses/xaddress.h>
+#include <config.h>
 #include <random.h>
+#include <script/standard.h>
+#include <span.h>
 #include <test/util/setup_common.h>
 #include <util/strencodings.h>
 #include <util/vector.h>
@@ -95,4 +98,100 @@ BOOST_AUTO_TEST_CASE(raw_encode_decode_fails) {
     }
 }
 
+BOOST_AUTO_TEST_CASE(parse_p2pkh_works) {
+    // First create an instance of an engine.
+    std::random_device rd;
+    auto gen = [&rd]() { return uint8_t(rd()); };
+
+    GlobalConfig config;
+    // Assert that parsing returns a correct CTxDestination for PKHash.
+    for (size_t i = 0; i < 10000; i++) {
+        std::vector<uint8_t> fakeKeyBytes(20);
+        std::generate(begin(fakeKeyBytes), end(fakeKeyBytes), gen);
+        CScript scriptPubKey = CScript() << OP_DUP << OP_HASH160
+                                         << ToByteVector(fakeKeyBytes)
+                                         << OP_EQUALVERIFY << OP_CHECKSIG;
+        const std::string encodedAddress = XAddress::Encode(XAddress::Content(
+            XAddress::TOKEN_NAME, XAddress::MAINNET, XAddress::SCRIPT_PUB_KEY,
+            ToByteVector(scriptPubKey)));
+        CTxDestination dest;
+        BOOST_CHECK_EQUAL(
+            XAddress::Parse(config.GetChainParams(), encodedAddress, dest),
+            true);
+        const PKHash pkDest = boost::get<PKHash>(dest);
+        BOOST_CHECK_EQUAL(MakeSpan(pkDest) == fakeKeyBytes, true);
+    }
+
+    // Assert that parsing returns a correct CTxDestination for ScriptHash.
+    for (size_t i = 0; i < 10000; i++) {
+        std::vector<uint8_t> fakeHashBytes(20);
+        std::generate(begin(fakeHashBytes), end(fakeHashBytes), gen);
+        CScript scriptPubKey =
+            CScript() << OP_HASH160 << ToByteVector(fakeHashBytes) << OP_EQUAL;
+        const std::string encodedAddress = XAddress::Encode(XAddress::Content(
+            XAddress::TOKEN_NAME, XAddress::MAINNET, XAddress::SCRIPT_PUB_KEY,
+            ToByteVector(scriptPubKey)));
+        CTxDestination dest;
+        BOOST_CHECK_EQUAL(
+            XAddress::Parse(config.GetChainParams(), encodedAddress, dest),
+            true);
+        const ScriptHash pkDest = boost::get<ScriptHash>(dest);
+        BOOST_CHECK_EQUAL(MakeSpan(pkDest) == fakeHashBytes, true);
+    }
+
+    // Check that parsing return CNoDestination for garbage.
+    for (size_t i = 0; i < 10000; i++) {
+        std::vector<uint8_t> garbledyGook(20);
+        std::generate(begin(garbledyGook), end(garbledyGook), gen);
+        const std::string encodedAddress = XAddress::Encode(
+            XAddress::Content(XAddress::TOKEN_NAME, XAddress::MAINNET,
+                              XAddress::SCRIPT_PUB_KEY, garbledyGook));
+        CTxDestination dest;
+        XAddress::Content parsedAddress;
+        BOOST_CHECK_EQUAL(
+            XAddress::Parse(config.GetChainParams(), encodedAddress, dest),
+            false);
+        BOOST_CHECK_NO_THROW(boost::get<CNoDestination>(dest));
+    }
+    // Check that parsing return CNoDestination for unknown type byte.
+    {
+        CTxDestination dest;
+        std::vector<uint8_t> fakeHashBytes(20);
+        std::generate(begin(fakeHashBytes), end(fakeHashBytes), gen);
+        CScript scriptPubKey =
+            CScript() << OP_HASH160 << ToByteVector(fakeHashBytes) << OP_EQUAL;
+
+        // Check that parsing return CNoDestination for unknown type byte.
+        BOOST_CHECK_EQUAL(
+            XAddress::Parse(
+                config.GetChainParams(),
+                XAddress::Encode(XAddress::Content(
+                    XAddress::TOKEN_NAME, XAddress::MAINNET,
+                    static_cast<XAddress::AddressType>(0xf0), fakeHashBytes)),
+                dest),
+            false);
+        BOOST_CHECK_NO_THROW(boost::get<CNoDestination>(dest));
+
+        // Check that parsing return CNoDestination for unknown token.
+        BOOST_CHECK_EQUAL(
+            XAddress::Parse(config.GetChainParams(),
+                            XAddress::Encode(XAddress::Content(
+                                "poop", XAddress::MAINNET,
+                                XAddress::SCRIPT_PUB_KEY, fakeHashBytes)),
+                            dest),
+            false);
+        BOOST_CHECK_NO_THROW(boost::get<CNoDestination>(dest));
+
+        // Check that parsing return CNoDestination for wrong network.
+        BOOST_CHECK_EQUAL(
+            XAddress::Parse(config.GetChainParams(),
+                            XAddress::Encode(XAddress::Content(
+                                XAddress::TOKEN_NAME, XAddress::TESTNET,
+                                XAddress::SCRIPT_PUB_KEY, fakeHashBytes)),
+                            dest),
+            false);
+        BOOST_CHECK_NO_THROW(boost::get<CNoDestination>(dest));
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
-- 
2.39.5 (Apple Git-154)

